cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -329,10 +480,9 @@ struct PrimitiveTakeImpl {
         if (block.popcount == block.length) {
           // Faster path: indices are not null but values may be
           for (int64_t i = 0; i < block.length; ++i) {
-            if (bit_util::GetBit(values_is_valid,
-                                 values_offset + indices_data[position])) {
+            if (values.IsValid(indices_data[position])) {
               // value is not null
-              out[position] = values_data[indices_data[position]];
+              out[position] = values.GetValue(indices_data[position]);
```

Comment:
I'll note that in the `IsValid` -> `GetValue` sequence the chunked resolution is called twice. The compiler might be able to optimize away the second call but that's not guaranteed.